system/lib/wasmfs/syscalls.cpp

```diff
@@ -513,4 +537,69 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
   }
   return __WASI_ERRNO_SUCCESS;
 }
+
+long __syscall_getdents64(long fd, long dirp, long count) {
+  auto openFile = wasmFS.getLockedFileTable()[fd];
+
+  if (!openFile) {
+    return -EBADF;
+  }
+  dirent* result = (dirent*)dirp;
+
+  // Check if the result buffer is too small.
+  if (count / sizeof(dirent) == 0) {
+    return -EINVAL;
+  }
+
+  auto file = openFile.locked().getFile();
+
+  auto directory = file->dynCast<Directory>();
+
+  if (!directory) {
+    return -ENOTDIR;
+  }
+
+  // Hold the locked directory for the duration of this operation.
+  auto lockedDir = directory->locked();
+
+  off_t bytesRead = 0;
+  // A directory's position corresponds to the index in its entries vector.
+  int index = openFile.locked().position();
+
+  // In the root directory, ".." refers to itself.
+  auto dotdot =
+    file == wasmFS.getRootDirectory() ? nullptr : lockedDir.getParent();
```

Comment:
The `..` for the root directory should be itself. But if the directory we are reading is unlinked, it's still possible for `dotdot` to be null. I'm not sure whether we should include the `..` entry in that case, or if we should, what value it should have. Could you do some experiments to find out?

Reply:
I think you can't unlink a directory. I tried it out with `rmdir`, but I you can't remove the upper directory (`..`) since it has children.